### PR TITLE
Fix tainted replicated motion

### LINKED ISSUE #12219: Problem in working with GPORCA, REPLICATED and SERIAL
### Greenplum version or build
PostgreSQL 9.4.24 (Greenplum Database 6.12.1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609, 64-bit compiled on Nov 24 2020 14:10:52

### OS version and uname -a
Linux gpdb-master-01 4.19.62-13 #1 SMP Mon Jul 29 17:22:59 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

### autoconf options used ( config.status --config )
'--build=x86_64-linux-gnu' '--prefix=/usr' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--sysconfdir=/etc' '--localstatedir=/var' '--disable-silent-rules' '--libdir=${prefix}/lib/x86_64-linux-gnu' '--libexecdir=${prefix}/lib/x86_64-linux-gnu' '--disable-maintainer-mode' '--disable-dependency-tracking' '--with-perl' '--with-python' '--without-gssapi' '--disable-gpcloud' '--enable-ic-proxy' '--with-libxml' '--with-system-tzdata=/usr/share/zoneinfo' '--prefix=/usr/local/gpdb' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-Bsymbolic-functions -Wl,-z,relro' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -fstack-protector-strong -Wformat -Werror=format-security'

### Actual behavior
```sql
SET optimizer = ON;
DROP TABLE IF EXISTS "tmp_tbl1";
CREATE TEMP TABLE "tmp_tbl1"
(
    "val1" VARCHAR,
    "val2" VARCHAR
)
DISTRIBUTED REPLICATED;

INSERT INTO "tmp_tbl1"
SELECT random()::text AS val1, random()::text AS val2
FROM generate_series(1, 10000);

SELECT count(*)
FROM "tmp_tbl1";
--10000

DROP TABLE IF EXISTS "tmp_tbl2";
CREATE TEMP TABLE "tmp_tbl2"
(
    "_etl_id" BIGSERIAL,
    "val1"    VARCHAR,
    "val2"    VARCHAR
)
DISTRIBUTED BY ("_etl_id")
;

INSERT INTO "tmp_tbl2" ("val1", "val2")
SELECT "val1", "val2"
FROM "tmp_tbl1";

SELECT count(*)
FROM "tmp_tbl2";
--10129
```

### PR DESCRIPTION
**Pull request by dgkimura**
In DXL to PlStmt translation ORCA decides whether to translate a motion node
into a redistribute motion or a result hash filter node. In the case of tainted
replicated the wrong translation can lead to incorrect results.

Following demonstrates the issue of result hash filter node:
```
CREATE TABLE replicated_table(v VARCHAR) DISTRIBUTED REPLICATED;
CREATE TABLE distributed_table(id BIGSERIAL, v VARCHAR) DISTRIBUTED BY (id);
EXPLAIN (COSTS OFF) INSERT INTO distributed_table (v) SELECT v FROM replicated_table;
                   QUERY PLAN
------------------------------------------------
 Insert on distributed_table
   ->  Result
         ->  Result
               ->  Seq Scan on replicated_table
 Optimizer: Pivotal Optimizer (GPORCA)
```

In this plan the replicated table is scanned on each segment and generates a
unique sequence value. Since the sequence value is unique across all segments
we now have a tainted replicated distribution. And because each tuple on all
segments is unique, we not guaranteed to generate a single row in the
replicated table that hashes to exactly one segment. Instead we can get missing
and even duplicate rows.

Fix is to force any motion produced because of a tainted replicated
distribution to translate into a redistribute motion during DXL to PlStmt
translation. Now we will generate the following plan:
```
                                  QUERY PLAN
------------------------------------------------------------------------------
 Insert on distributed_table
   ->  Result
         ->  Redistribute Motion 1:3  (slice1; segments: 1)
               Hash Key: (nextval('distributed_table__etl_id_seq'::regclass))
               ->  Seq Scan on replicated_table
 Optimizer: Pivotal Optimizer (GPORCA)
```

Fixes #12219

### CODE REVIEW

File: src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 1]
```diff
@@ -7550,7 +7550,8 @@ CTranslatorExprToDXL::GetInputSegIdsArray(CExpression *pexprMotion)
 		return pdrgpi;
 	}
 
-	if (CUtils::FDuplicateHazardMotion(pexprMotion))
+	if (CUtils::FDuplicateHazardMotion(pexprMotion) ||
```
Comment (bhuvnesh2703) [inline src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:7553]: Let's add some comment and try if we can make it more intuitive..

**Review by bhuvnesh2703 — APPROVED** [review]
LGTM, but please fix the comment.

File: src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp (hunk 1 above)
Comment (dgkimura) [reply src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:7558]: Added comments to help make it more intuitive.

File: src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 2]
```diff
@@ -7550,10 +7555,16 @@ CTranslatorExprToDXL::GetInputSegIdsArray(CExpression *pexprMotion)
...
+		CDistributionSpec::EdtTaintedReplicated == pds->Edt())
 	{
 		// if Motion is duplicate-hazard, we have to read from one input segment
 		// to avoid generating duplicate values
+		//
+		// Additionally, if the child distribution is tainted-replicated, the
+		// motion (which cannot be a result hash filter node) will read the
```
Comment (bhuvnesh2703) [inline src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:7565]: can we just add a simple example here, for ex: limit where the scan on replicated table can give different results on different segments, so we can't use hash filter to filter out the rows as the data is not same across the segments
seg0 -> limit 1 can give 0
seg1 -> limit 1 can give 1

also minor rephrase `input from one segment in order to ensure that data is consistent after bring read from operator delivering tainted replication`

Comment (dgkimura) [reply src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:7586]: Done.  Added an example where CUtils::FDuplicateHazardMotion is used.

**Comment by dgkimura** [comment]
Merged. Thanks for review!

**Pull request merged**